pkg/controller/kinesis_trigger_controller.go

```diff
@@ -0,0 +1,353 @@
+}
+
+func (c *KinesisTriggerController) waitForCacheSync(stopCh <-chan struct{}) bool {
+	if !cache.WaitForCacheSync(stopCh, c.kinesisInformer.Informer().HasSynced, c.functionInformer.Informer().HasSynced) {
+		utilruntime.HandleError(fmt.Errorf("Timed out waiting for caches required for AWS Kinesis triggers controller to sync;"))
+		return false
+	}
+	c.logger.Info(" AWS Kinesis Trigger controller caches are synced and ready")
+	return true
+}
+
+func (c *KinesisTriggerController) runWorker() {
+	for c.processNextItem() {
+		// continue looping
+	}
+}
+
+func (c *KinesisTriggerController) processNextItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncKinesisTrigger(key.(string))
+	if err == nil {
+		// No error, reset the ratelimit counters
+		c.queue.Forget(key)
+	} else if c.queue.NumRequeues(key) < triggerMaxRetries {
+		c.logger.Errorf("Error processing %s (will retry): %v", key, err)
+		c.queue.AddRateLimited(key)
+	} else {
+		// err != nil and too many retries
+		c.logger.Errorf("Error processing %s (giving up): %v", key, err)
+		c.queue.Forget(key)
+		utilruntime.HandleError(err)
+	}
+
+	return true
+}
+
+func (c *KinesisTriggerController) syncKinesisTrigger(key string) error {
+	c.logger.Infof("Processing update to AWS Kinesis Trigger: %s", key)
+	ns, triggerObjName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	obj, exists, err := c.kinesisInformer.Informer().GetIndexer().GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("Error fetching object with key %s from store: %v", key, err)
+	}
+
+	// this is an update when Kinesis trigger API object is actually deleted, we dont need to process anything here
+	if !exists {
+		c.logger.Infof("Kinesis Trigger %s not found in the cache, ignoring the deletion update", key)
+		return nil
+	}
+
+	triggerObj := obj.(*kubelessApi.KinesisTrigger)
+
+	if triggerObj.Spec.Stream == "" {
+		return errors.New("Kinesis Trigger Stream can't be empty. Please check the Kinesis trigger object %s" + key)
+	}
+	if triggerObj.Spec.Region == "" {
+		return errors.New("Kinesis Trigger region can't be empty. Please check the Kinesis trigger object %s" + key)
+	}
+	if triggerObj.Spec.ShardID == "" {
+		return errors.New("Kinesis Trigger shard-id can't be empty. Please check the Kinesis trigger object %s" + key)
+	}
+	if triggerObj.Spec.Secret == "" {
+		return errors.New("Kinesis Trigger secret can't be empty. Please check the Kinesis trigger object %s" + key)
+	}
+	_, err = utils.GetFunctionCustomResource(c.kubelessclient, triggerObj.Spec.FunctionName, ns)
+	if err != nil {
+		return fmt.Errorf("Unable to find Function %s in namespace %s. Error %v", triggerObj.Spec.FunctionName, ns, err)
+	}
+	_, err = c.kubernetesClient.Core().Secrets(ns).Get(triggerObj.Spec.Secret, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("Unable to find secret %s in namespace %s. Error %v", triggerObj.Spec.Secret, ns, err)
```

Comment:
this two checks should be after checking if the trigger is being deleted (for the case in which the user deletes first the function/secret and then the trigger), isn't it?

Reply:
Agree. Moved the checks after the check to find trigger is getting deleted.